Fixed many python compatiblity issues

kochelmonster opened the pull request:
- relative import
- metaclass
- descriptor protocol
- id, frozenset, getattr
- missing modules
- type and super

kochelmonster committed:
enhanced relative import "from ..foo.bar import a" + support of type with 3 arguments.

kochelmonster committed:
implemented __init__ call of meta class.

kochelmonster committed:
extended core with id, frozenset and getattr with default argument,

kochelmonster committed:
enhanced inspect.js with getmro.

kochelmonster committed:
support of descriptor protocol.

kochelmonster committed:
new weakref module.

kochelmonster committed:
new operator module.

kochelmonster committed:
new functools module.

kochelmonster committed:
new collections module.

kochelmonster committed:
new bisect module.

kochelmonster committed:
type(function) returns a function type.
+ string.lstrip, string.rstrip with arguments.

kochelmonster committed:
added RuntimeError.

kochelmonster committed:
fixed bug in __getattr__

kochelmonster committed:
fixed relative import of modules "from . import module" + tests for relative import.

kochelmonster committed:
minor fixes.

kochelmonster committed:
"$" joins for command line arguments do not work in linux -> allow multiple arguments.

kochelmonster committed:
fixes for strict mode.

kochelmonster committed:
functools update_wrapper

kochelmonster committed:
fixed operator.itemgetter

kochelmonster committed:
__super__ now calls the correct mro.

kochelmonster committed:
enhanced tests for new __super__

kochelmonster committed:
Activate method memoizing + keep attributes of method.

kochelmonster committed:
Activated the method memoizing + keep attrbutes methods.

kochelmonster committed:
Merge branch 'master' of https://github.com/kochelmonster/Transcrypt

kochelmonster committed:
super handling: correct initialization of classes if __init__ is missing. Improved __super__.

kochelmonster committed:
Another fix for super

kochelmonster committed:
Improved metaclass handling.

kochelmonster committed:
Another fix for super.

kochelmonster committed:
Merge remote-tracking branch 'upstream/master'

kochelmonster committed:
Merge remote-tracking branch 'upstream/master'

JennaSys commented:
Thank you for the contribution. There is a lot in here that I'll have to dig through. TBH, I'll likely end up having to cherry pick some of the updates rather than do one massive merge. But, at least based on the commit messages, I'm looking forward to adding some of these features.

Also just so you know, I'm currently in the process of updating the Transcrypt compiler to be compatible with Python 3.13 and any new features will need to be added after that is released.

kochelmonster committed:
Fixed assignment bug in local scope:

def count_100(index):
    for j in range(100):
        index = index + 1
    return index

Generated wrong to:

export var count_100 = function (index) {
	for (var j = 0; j < 100; j++) {
		var index = index + 1;  // var is wrong
	}
	return index;
};

------

def count_101(index):
    idx = index
    for j in range(101):
        idx = idx + 1
    return idx

Generated wrong to:

export var count_101 = function (index) {
	var idx = index;
	for (var j = 0; j < 101; j++) {
		var idx = idx + 1;   // var is wrong
	}
	return idx;
};
